### PR TITLE
Feature/eip1 4461 approved postal application database entity and repository components

### PR DESCRIPTION
**Pull request by spalamthodi**
https://technologyprogramme.atlassian.net/browse/EIP1-4461
Approved Postal Vote Application Entity / Repository

### CODE REVIEW

File: src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    var id: UUID? = null,
+
+    @NotNull
+    val street: String,
+
+    var property: String? = null,
+    var locality: String? = null,
+    var town: String? = null,
+    var area: String? = null,
+    @NotNull
```
Comment (gurpiarbassi) [inline src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt:34]: use @field:NotNull on javax annotations.
I have found that these do not get carried to the java byte code unless you do this.

Will also need tests to validate this. We are doing this at the moment in OE (Shumail has a PR on it at the moment)

Comment (spalamthodi) [reply src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt]: Thanks Gurps, I have updated all the beans

Comment (gurpiarbassi) [reply src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt]: tests needed to prove validation is firing. See current OE PR for sample

File: src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/ApplicantDetails.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import javax.persistence.JoinColumn
+import javax.persistence.OneToOne
+import javax.validation.constraints.NotNull
+
+@Embeddable
+class ApplicantDetails(
+    @field:NotNull
+    val firstName: String,
+    var middleNames: String? = null,
+    @field:NotNull
```
Comment (gurpiarbassi) [inline src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/ApplicantDetails.kt:18]: would be more clearer if we have new lines e.g.

```
@field:NotNull
val firstName: String,

var middleNames: String? = null,
```

File: src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+        // When
+        approvedPostalVoteApplicationRepository.saveAndFlush(approvedPostalVoteApplication)
+
+        // Then
+        val savedApplication =
+            approvedPostalVoteApplicationRepository.findById(approvedPostalVoteApplication.applicationId).get()
+
+        assertThat(savedApplication).usingRecursiveComparison().isEqualTo(approvedPostalVoteApplication)
+    }
```
Comment (inderjitvirdi) [inline src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt:24]: Is it worth adding some non happy path tests e.g.  for testing the constraint  validation is working?

Comment (gurpiarbassi) [reply src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt:24]: yeh suggest doing that as unit test. See OE as example

File: src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import uk.gov.dluhc.emsintegrationapi.testsupport.testdata.buildApprovedPostalApplication
+
+class ApprovedPostalVoteApplicationRepositoryTest : AbstractRepositoryTest() {
+    @Autowired
+    private lateinit var approvedPostalVoteApplicationRepository: ApprovedPostalVoteApplicationRepository
+    @Test
```
Comment (gurpiarbassi) [inline src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt:11]: could do with new line here

File: src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.dluhc.emsintegrationapi.database.repository
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import uk.gov.dluhc.emsintegrationapi.testsupport.testdata.buildApprovedPostalApplication
+
+class ApprovedPostalVoteApplicationRepositoryTest : AbstractRepositoryTest() {
```
Comment (gurpiarbassi) [inline src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt:8]: should the name of this class and its parent have the word "Integration" at least to tell us it's a integration test?

Comment (spalamthodi) [reply src/test/kotlin/uk/gov/dluhc/emsintegrationapi/database/repository/ApprovedPostalVoteApplicationRepositoryTest.kt]: Done

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 6]
```diff
@@ -2,6 +2,6 @@ databaseChangeLog:
   - include:
       file: /db/changelog/create/0001_address.sql
   - include:
-      file: /db/changelog/create/0002_approved_postal_application.sql
+      file: /db/changelog/create/0002_approved_postal_vote_application.sql
   - include:
-      file: /db/changelog/create/0003_approved_proxy_application.sql
+      file: /db/changelog/create/0003_approved_proxy_vote_application.sql
```
Comment (kristerbone) [inline src/main/resources/db/changelog/db.changelog-master.yaml:7]: Do we maybe want to just call these xxxx_applications - we know we'll be sending both approved and rejected and could a more painful change later.

File: src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    @field:Size(max = 255)
+    var area: String? = null,
+
+    @field:Size(max = 10)
+    val postcode: String,
+
+    @field:Size(max = 12)
+    var uprn: String? = null,
+
+    @CreationTimestamp
```
Comment (gurpiarbassi) [inline src/main/kotlin/uk/gov/dluhc/emsintegrationapi/database/entity/Address.kt:48]: interested to see if these timestamps are actually being persisted or not without the entity listener declaration.

**Review by gurpiarbassi — APPROVED** [review]

**Pull request merged**